registry/storage/driver/middleware/cloudfront/s3filter_test.go

```diff
@@ -0,0 +1,374 @@
+	if updateCount < 4 {
+		t.Errorf("Update should have been called at least 4 times, actual=%d", updateCount)
+	}
+}
+
+func TestClientCheck(t *testing.T) {
+	empty := context.TODO()
+
+	makeContext := func(useragent string) context.Context {
+		// makeContext returns a context with an http request with
+		// the given user agent
+		req := &http.Request{
+			Header: map[string][]string{
+				"User-Agent": {useragent},
+			},
+		}
+
+		return dcontext.WithRequest(empty, req)
+	}
+	cases := []struct {
+		Context  context.Context
+		Expected bool
+	}{
+		{Context: empty, Expected: false},
+		{Context: makeContext(""), Expected: false},
+		{Context: makeContext("docker/"), Expected: false},
+		{Context: makeContext("docker/0.7.0"), Expected: false},
+		{Context: makeContext("docker/1.5.0"), Expected: false},
+		{Context: makeContext("docker/1.6.0"), Expected: true},
+		{Context: makeContext("docker/1.11.0"), Expected: true},
+		{Context: makeContext("docker/17.03.0"), Expected: true},
+		{Context: makeContext("docker/17.03.0-ce"), Expected: true},
+		{Context: makeContext("docker/17.03.1-ee-2 go/go1.7.5"), Expected: true},
+		{Context: makeContext("curl/7.9.8"), Expected: false},
+	}
+
+	for _, testCase := range cases {
+		t.Run(fmt.Sprintf("Client useragent = %v", testCase.Context.Value("http.request.useragent")), func(t *testing.T) {
+			AssertEqual(t, testCase.Expected, clientDropsAuthHeader(testCase.Context))
+		})
+	}
+}
+
+// populate ips with a number of different ipv4 and ipv6 networks, for the purposes
+// of benchmarking contains() performance.
+func populateRandomNetworks(b *testing.B, ips *aWSIPs, ipv4Count, ipv6Count int) {
+	generateNetworks := func(dest *[]net.IPNet, bytes int, count int) {
+		for i := 0; i < count; i++ {
+			ip := make([]byte, bytes)
+			_, err := rand.Read(ip)
+			if err != nil {
+				b.Fatalf("Failed to generate network for test : %s", err.Error())
+			}
+			mask := make([]byte, bytes)
+			for i := 0; i < bytes; i++ {
+				mask[i] = 0xff
+			}
+			*dest = append(*dest, net.IPNet{
+				IP:   ip,
+				Mask: mask,
+			})
+		}
+	}
+
+	generateNetworks(&ips.ipv4, 4, ipv4Count)
+	generateNetworks(&ips.ipv6, 16, ipv6Count)
+}
+
+func BenchmarkContainsRandom(b *testing.B) {
+	// Generate a random network configuration, of size comparable to
+	// aws official networks list
+	// curl -s https://ip-ranges.amazonaws.com/ip-ranges.json | jq '.prefixes | length'
+	// 856
+	numNetworksPerType := 1000 // keep in sync with the above
+	// intentionally skip constructor when creating aWSIPs, to avoid updater routine.
+	// This benchmark is only concerned with contains() performance.
+	awsIPs := aWSIPs{}
+	populateRandomNetworks(b, &awsIPs, numNetworksPerType, numNetworksPerType)
+
+	ipv4 := make([]byte, 4)
+	ipv6 := make([]byte, 16)
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		rand.Read(ipv4)
+		rand.Read(ipv6)
+		awsIPs.contains(ipv4)
+		awsIPs.contains(ipv6)
+	}
+}
+
+func BenchmarkContainsProd(b *testing.B) {
```

Comment:
What are some numbers for this benchmark? Right now, matching is based on a loop, which is fine for small N. However, it looks like there are on the order of 1000 ipv4 and ipv6 addresses in the production structure.